docs/source/extension/plugin_system.rst

```diff
@@ -0,0 +1,179 @@
+.. Copyright (c) Jupyter Development Team.
+.. Distributed under the terms of the Modified BSD License.
+
+The JupyterLab Plugin System
+============================
+
+JupyterLab's plugin system is designed so that plugins can depend on and
+reuse features from one another. A key part of this approach is JupyterLab's
+**provider-consumer pattern**.
+
+The Provider-Consumer Pattern
```

Comment:
I appreciate the need for naming things, but I wonder if this does not make the document less approachable. This section has a substantial overlap with [Plugins Interacting with Each Other](https://github.com/jupyterlab/jupyterlab/blob/main/docs/source/extension/extension_dev.rst#plugins-interacting-with-each-other) which is much friendlier title for a section in my opinion.

Reply:
I agree that some renaming/a friendlier title would be good here, though still think it's important to keep the term in there so people have something to point toward that describes the paradigm consicely.

Reply:
> substantial overlap with Plugins Interacting with Each Other

Also wondering why we need to re-explain the concepts already covered in [Plugins Interacting with Each Other](https://github.com/jupyterlab/jupyterlab/blob/main/docs/source/extension/extension_dev.rst#plugins-interacting-with-each-other)? That existing section is maybe a bit longer than the new one proposed in this PR, but still concise.